Add arrow and tab navigation in FlatList.

rectified95 opened the pull request:
Fixes #2107
Aligning RNW with native XAML behavior: using arrows will move focus between list entries, and tab will move navigation in/out of the list to adjacent UI components.
The focus seeking code was developed by @lyahdav (linked here #6293). I'm adapting it by registering listeners to the Preview* event analogs, in order to intercept them before they get marked as handled by XAML while bubbling up. 

Also, FlatList now "remembers" the index of the last focused element before focus left the list, also in line with XAML.

Since RNW uses the same native XAML ScrollViewer to implement both the RN FlatList and ScrollView components, we need a way of telling the ScrollViewManager which component it is rendering and conditionally register the event handlers appropriate for FlatList; ScrollView behavior should remain unaltered, in line with how native UWP apps act. To accomplish this I'm introducing a prop - `isFlatList` - that VirtualizedList passed down to ScrollView. This requires forking the component to make that tiny change. I'm very much open to suggestions how to do it differently.


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6565)

acoates-ms commented:
I'm not a fan of this.  We are changing core components, to handle very specific things, and doing in a way that removes control from JS rather than adds capabilities to JS.

acoates-ms commented:
The logic is all in native tightly coupled to the scrollview.  From the JS side this doesn't help users of the platform solve similar problems (such as radio group).

The point of FocusZone was to take this logic and put it in a discrete reusable block that can be used in many interesting ways across many controls.

rectified95 commented:
I see your point. @harinikmsft suggested that since this fixes the broken FlatList behavior, FocusZone becomes less of a priority, so I removed the FocusZone component from this PR.

I could create a separate PR or modify this one and readd FocuZone. I've tested it locally for a group of Buttons and it worked. Still, wrapping FlatList in FocusZone as LIron's code initially suggested wouldn't fix FlatList navigation, but then I could just have it call into portions of the FocusZone code with event registration. Would that be better? I think I would still need the JS prop for Flatlist, though.

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -10,6 +10,9 @@
 #include "Impl/ScrollViewUWPImplementation.h"
 #include "ScrollViewManager.h"
 
+#include <winrt/Windows.UI.Core.h>
+#include <winrt/Windows.UI.Xaml.Input.h>
```
these should be using the straddle headers, not the cppwinrt ones
(#include "UI.Core.h" etc.)

rectified95 replied:
What are straddle headers and why should I use them here?

asklar replied:
the headers that allow us to reason about a xaml type regardless of whether it's in UWP XAML or WinUI 3.
You should use them here so that you don't break WinUI 3 RNW builds.

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -146,6 +221,24 @@ void ScrollViewShadowNode::updateProperties(winrt::Microsoft::ReactNative::JSVal
+    } else if (propertyName == "isFlatList") {
+      const auto [valid, isFlatList] = getPropertyAndValidity(propertyValue, true);
+      if (valid && isFlatList) {
+        // We use the XAMl ScrollViewer for both the RN ScrollView and FlatList - need arrow navigation of FlatList
```
```suggestion
        // We use the XAML ScrollViewer for both the RN ScrollView and FlatList - need arrow navigation of FlatList only
```

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -146,6 +221,24 @@ void ScrollViewShadowNode::updateProperties(winrt::Microsoft::ReactNative::JSVal
+      if (valid && isFlatList) {
+        // We use the XAMl ScrollViewer for both the RN ScrollView and FlatList - need arrow navigation of FlatList
+        // entries.
+        m_PreviewKeyDownRevoker = scrollViewer.PreviewKeyDown(
+            winrt::auto_revoke, [=](const winrt::IInspectable &sender, const xaml::Input::KeyRoutedEventArgs &e) {
+              OnPreviewKeyDown(sender, e);
+            });
+        m_GettingFocusRevoker = scrollViewer.GettingFocus(
+            winrt::auto_revoke, [=](const winrt::IInspectable &sender, const xaml::Input::GettingFocusEventArgs &e) {
+              OnGettingFocus(sender, e);
+            });
+        m_LosingFocusRevoker = scrollViewer.LosingFocus(
+            winrt::auto_revoke, [=](const winrt::IInspectable &sender, const xaml::Input::LosingFocusEventArgs &e) {
+              OnLosingFocus(sender, e);
```
you can omit the types of the lambdas and just do `const auto&` since it's not important you're just forwarding

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -121,6 +131,71 @@ void ScrollViewShadowNode::createView() {
+    findNextElementOptions.ExclusionRect(exclusionRect);
+
+    auto nextElement = xaml::Input::FocusManager::FindNextElement(
+        isShiftDown ? xaml::Input::FocusNavigationDirection::Up : xaml::Input::FocusNavigationDirection::Down,
```
Please check the result of the WinUI 3 build (which will currently not fail the PR build if it fails)

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -121,6 +131,71 @@ void ScrollViewShadowNode::createView() {
+      xaml::Input::FocusManager::TryMoveFocus(xaml::Input::FocusNavigationDirection::Next, findNextElementOptions);
+      e.Handled(true);
+    }
+  } else if (e.Key() == winrt::Windows::System::VirtualKey::Tab) {
```
this may need to use the straddle VKs instead of Windows::System

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -121,6 +131,71 @@ void ScrollViewShadowNode::createView() {
+void ScrollViewShadowNode::OnPreviewKeyDown(
+    const winrt::IInspectable &sender,
+    const xaml::Input::KeyRoutedEventArgs &e) {
+  auto senderFrameworkElement = sender.try_as<winrt::FrameworkElement>();
```
```suggestion
  auto senderFrameworkElement = sender.try_as<xaml::FrameworkElement>();
```

asklar reviewed (changes requested):
please make sure the new code is using the straddled types instead of raw windows xaml

acoates-ms commented:
Would it work if the flatlist created the FocusZone within the scrollview from the JS side?

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -121,6 +131,71 @@ void ScrollViewShadowNode::createView() {
+      e.Handled(true);
+    }
+  } else if (e.Key() == winrt::Windows::System::VirtualKey::Tab) {
+    auto const &coreWindow = winrt::CoreWindow::GetForCurrentThread();
```
needs an alternative for XAML Islands

lyahdav replied:
When will this break in Islands? We're using this code now, but only in 1 window.

asklar replied:
more than one [core]window per thread

lyahdav replied:
I see `winrt::CoreWindow::GetForCurrentThread()` being used in several places in this repo. Are those all potential bugs then?

asklar replied:
Yes, any getforcurrent {thread, view, window} are potential bugs. I had done a pass through to ensure we had meaningful fallbacks. At the same time, xaml islands isn't a scenario we _officially_ support for RNW (https://github.com/microsoft/react-native-windows-samples/issues/267), but if there is a scenario you are blocked on and there is a simple fix on the RNW side, we'd take it :)

lyahdav commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -121,6 +131,71 @@ void ScrollViewShadowNode::createView() {
+    auto nextElement = xaml::Input::FocusManager::FindNextElement(
+        isShiftDown ? xaml::Input::FocusNavigationDirection::Up : xaml::Input::FocusNavigationDirection::Down,
+        findNextElementOptions);
+    xaml::Input::FocusManager::TryFocusAsync(nextElement, winrt::FocusState::Programmatic);
```
I found a bug here: this should be wrapped in a `if (nextElement) { ... }`. Otherwise it will crash if it can't find a next element.

rectified95 replied:
Good catch, I ran into this yesterday too!

lyahdav commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -121,6 +131,71 @@ void ScrollViewShadowNode::createView() {
+    auto exclusionRect = winrt::Rect(
+        anchorTopLeftConverted.X,
+        anchorTopLeftConverted.Y,
+        static_cast<float>(senderFrameworkElement.Width()),
+        static_cast<float>(senderFrameworkElement.Height()));
```
I found a bug here: We should be using `ActualWidth` and `ActualHeight`

rectified95 replied:
What difference have you noticed it made?

lyahdav replied:
In cases where the FocusZoneView didn't have an explicit width/height it wouldn't return the correct next element. The Width/Height properties are set only if you explicitly size a control in Xaml. You should always use the `Actual` props because they will give you the computed size of a control. @asklar helped me debug this.

lyahdav commented:
> wrapping FlatList in FocusZone as LIron's code initially suggested wouldn't fix FlatList navigation, but then I could just have it call into portions of the FocusZone code with event registration. Would that be better? I think I would still need the JS prop for Flatlist, though.

If you wrap FlatList in a FocusZone what issues does FlatList navigation have? Also, I never tested this approach with a FlatList where each row has more than one focusable control. You might want to see what Xaml does in that case and see if we can match it.

asklar commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -121,6 +131,71 @@ void ScrollViewShadowNode::createView() {
+      xaml::Input::FocusManager::GetFocusedElement(xamlRoot).try_as<winrt::Windows::UI::Xaml::DependencyObject>();
+  auto findNextElementOptions = xaml::Input::FindNextElementOptions();
+  findNextElementOptions.SearchRoot(xamlRoot.Content());
+  auto previousKey = m_isHorizontal ? winrt::Windows::System::VirtualKey::Left : winrt::Windows::System::VirtualKey::Up;
+  auto nextKey = m_isHorizontal ? winrt::Windows::System::VirtualKey::Right : winrt::Windows::System::VirtualKey::Down;
```
is this correct for RTL languages?

lyahdav replied:
Good question, I didn't test RTL

rectified95 commented:
@lyahdav Apologies - a while back I had issues wrapping a Flatlist in the FocusZone and so for the longest time thought it didn't work - I may have used a non-focusable element as list member that caused it.. 
Just tried this with a list of buttons and it works like a charm!

Will also double-check the scenario with multiple focusable elements per entry.

NickGerleman commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -146,6 +221,24 @@ void ScrollViewShadowNode::updateProperties(winrt::Microsoft::ReactNative::JSVal
+        // We use the XAMl ScrollViewer for both the RN ScrollView and FlatList - need arrow navigation of FlatList
+        // entries.
+        m_PreviewKeyDownRevoker = scrollViewer.PreviewKeyDown(
+            winrt::auto_revoke, [=](const winrt::IInspectable &sender, const xaml::Input::KeyRoutedEventArgs &e) {
```
Implicit capture by value (the `[=]`) should generally be avoided because it makes it too easy to introduce subtle memory safety issues or ref cycles. Reminded of this after a PR came in today for this happening (https://github.com/microsoft/react-native-windows/pull/6592) 😀.

So, for this specific case, we're relying on the raw `this` ptr inside the lambda, and the equivalent would be to replace `[=]` with `[this]`. Capturing a raw pointer in an async lambda generally best to avoid compared to something with weak owership, but is safe here because the revoker ensures the callback will be unregistered before the object is destroyed.

NickGerleman reviewed (commented):
Left a tiny nit. There are folks on this review who will have a lot more context on the right overall approach here than I do.

rectified95 commented:
Restoring FocusZone component. Wrapping a FlatList in it does provide arrow navigation, but compared to the previous approach of modifying the ScrollViewManager, tabbing behavior is not fully correct: instead of moving focus out of the list to another component, it changes it to an element in the list lower that the currently visible portion (as if PageDown had been pressed). Also, the caching of the last focused element is partly broken in this approach.
Need to debug further to understand why that is.
Addressed other comments, and moved updated code to FocusZoneViewManager.cpp

asklar commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.cpp:
```diff
@@ -0,0 +1,151 @@
+namespace winrt {
+using namespace Windows::Foundation;
+using namespace Windows::UI;
+using namespace Windows::UI::Xaml;
+using namespace Windows::UI::Xaml::Controls;
+using namespace Windows::UI::Xaml::Controls::Primitives;
+using namespace Windows::UI::Xaml::Media;
```
don't use Windows::UI::Xaml directly, only reference the aliased location (xaml, xaml::Controls, etc.), otherwise this will break WinUI 3 builds of RNW.

asklar replied:
also no need to re-alias these into winrt namespace

rectified95 replied:
Sorry, I missed these when updating the namespaces around method calls below.

asklar commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.cpp:
```diff
@@ -0,0 +1,151 @@
+    const auto &propertyValue = pair.second;
+
+    if (propertyName == "focusDirection") {
+      if (propertyValue.Type() == winrt::Microsoft::ReactNative::JSValueType::String) {
```
IMO it's better to expect a string than to no-op if we get passed something else. At least this way you'll get a crash and you can fix the bug, rather than trying to figure out why it didn't work as you expected

rectified95 replied:
I've looked at the other view managers and they seem to assign some default on failure so I've assigned false here.

asklar commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.cpp:
```diff
@@ -0,0 +1,151 @@
+      auto exclusionRect = winrt::Rect(
+          anchorTopLeftConverted.X,
+          anchorTopLeftConverted.Y,
+          static_cast<float>(senderFrameworkElement.Width()),
+          static_cast<float>(senderFrameworkElement.Height()));
```
these can be NaN if this happens before layout, use `ActualWidth` and `ActualHeight` instead

rectified95 replied:
Got lost after applying the change to ScrollViewManager.. fixing in next iteration.

asklar commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.cpp:
```diff
@@ -0,0 +1,151 @@
+void FocusZoneViewShadowNode::OnLosingFocus(
+    const winrt::IInspectable &sender,
+    const xaml::Input::LosingFocusEventArgs &e) {
+  m_lastElement = e.OldFocusedElement();
```
do you need to set `Handled`?

rectified95 replied:
Not sure about this one - I'm thinking that there might be actions related to focus visual taken by the system, so I'd pass this one up the routed chain. What do you think?

asklar commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.h:
```diff
@@ -0,0 +1,53 @@
+  void OnLosingFocus(const winrt::IInspectable &sender, const xaml::Input::LosingFocusEventArgs &e);
+
+  bool m_isHorizontal = false;
+  winrt::weak_ref<winrt::DependencyObject> m_lastElement = nullptr;
```
```suggestion
  winrt::weak_ref<xaml::DependencyObject> m_lastElement = nullptr;
```

asklar commented on vnext/Microsoft.ReactNative/Views/ViewViewManager.h:
```diff
@@ -3,12 +3,244 @@
+#include <inspectable.h>
+#include <unicode.h>
+#include <winrt/Windows.System.h>
+#include <winrt/Windows.UI.Xaml.Interop.h>
```
why is this needed?

asklar commented on vnext/src/Libraries/Components/Pressable/Pressable.windows.js:
```diff
@@ -187,14 +187,16 @@ function Pressable(props: Props, forwardedRef): React.Node {
   // [Windows
   const _onBlur = (event: BlurEvent) => {
-    TextInputState.blurInput(viewRef.current);
+    //TextInputState.blurInput(viewRef.current);
```
remove if not needed (and the console.log). Same below.

asklar commented on vnext/Microsoft.ReactNative/Views/ViewViewManager.cpp:
```diff
@@ -31,230 +31,6 @@
 
 namespace Microsoft::ReactNative {
 
-// ViewShadowNode
```
move the class declaration to the header, but keep the methods' implementation in the cpp file

asklar commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace Microsoft::ReactNative {
+
+class FocusZoneViewShadowNode : public ViewShadowNode {
```
this can be final

lyahdav commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  auto panel = GetViewPanel();
+  m_panelKeyDownRevoker =
+      panel.PreviewKeyDown(winrt::auto_revoke, [=](const auto &sender, const auto &e) { OnKeyDown(sender, e); });
```
Just for my understanding, why is `PreviewKeyDown` preferred over `KeyDown`? I can't tell the difference from the docs.

rectified95 replied:
My initial implementation registered the events inside ScrollViewManager - using the bubbled rather than the routed event resulted in both the XAML element scrolling the content, and our manager moving focus to the next element (e.handled was not getting set early enough). Given that the current implementation doesn't modify the ScrollViewManager but rather does use the FocusZone component, the preview event might no longer be necessary.

rectified95 replied:
The Tab behavior is still broken however in this implementation - pressing tab while there are more elements than currently visible in the list acts as if PgDown had been pressed, not moving focus outside the FlatList. I'm working with Keith and others on figuring out how to do this.

lyahdav commented on packages/playground/Samples/focusZone.tsx:
```diff
@@ -0,0 +1,168 @@
+  FlatList,
+} from 'react-native';
+
+const FocusZoneView = requireNativeComponent('FocusZoneView');
```
Seems to me we should have a FocusZoneView.tsx as part of this PR so end users don't need to use `requireNativeComponent`?

rectified95 replied:
True, the current state of the PR is straddled between two implementations, so I may have missed some bits when moving between the two. Good call.

lyahdav commented on vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:
```diff
@@ -53,6 +58,7 @@ class ScrollViewShadowNode : public ShadowNodeBase {
   bool m_dismissKeyboardOnDrag = false;
 
   std::shared_ptr<SIPEventHandler> m_SIPEventHandler;
+  winrt::weak_ref<winrt::DependencyObject> m_lastSelectedItem = nullptr;
```
Are the changes in this file still needed?

rectified95 replied:
See comment above.

lyahdav commented on vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.cpp:
```diff
@@ -0,0 +1,148 @@
+      }
+    }
+  }
+}
```
Sorry, I found another bug here. We need to call `Super::updateProperties(std::move(props));` so that View props are updated properly.
